src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableDefaultConverter.cs

```diff
@@ -18,40 +18,48 @@ protected override bool OnWriteResume(Utf8JsonWriter writer, TCollection value,
         {
             Debug.Assert(value is not null);
 
-            IEnumerator<TElement> enumerator;
-            if (state.Current.CollectionEnumerator == null)
+            IEnumerator<TElement>? enumerator = default;
+            try
             {
-                enumerator = value.GetEnumerator();
-                if (!enumerator.MoveNext())
+                if (state.Current.CollectionEnumerator == null)
                 {
-                    enumerator.Dispose();
-                    return true;
+                    enumerator = value.GetEnumerator();
+                    if (!enumerator.MoveNext())
+                    {
+                        enumerator.Dispose();
+                        return true;
+                    }
                 }
-            }
-            else
-            {
-                Debug.Assert(state.Current.CollectionEnumerator is IEnumerator<TElement>);
-                enumerator = (IEnumerator<TElement>)state.Current.CollectionEnumerator;
-            }
-
-            JsonConverter<TElement> converter = GetElementConverter(ref state);
-            do
-            {
-                if (ShouldFlush(writer, ref state))
+                else
                 {
-                    state.Current.CollectionEnumerator = enumerator;
-                    return false;
+                    Debug.Assert(state.Current.CollectionEnumerator is IEnumerator<TElement>);
+                    enumerator = (IEnumerator<TElement>)state.Current.CollectionEnumerator;
                 }
 
-                TElement element = enumerator.Current;
-                if (!converter.TryWrite(writer, element, options, ref state))
+                JsonConverter<TElement> converter = GetElementConverter(ref state);
+                do
                 {
-                    state.Current.CollectionEnumerator = enumerator;
-                    return false;
-                }
+                    if (ShouldFlush(writer, ref state))
+                    {
+                        state.Current.CollectionEnumerator = enumerator;
+                        return false;
+                    }
+
+                    TElement element = enumerator.Current;
+                    if (!converter.TryWrite(writer, element, options, ref state))
+                    {
+                        state.Current.CollectionEnumerator = enumerator;
+                        return false;
+                    }
 
-                state.Current.EndCollectionElement();
-            } while (enumerator.MoveNext());
+                    state.Current.EndCollectionElement();
+                } while (enumerator.MoveNext());
+            }
+            catch
```

Comment:
Could this also be done in a finally to reduce code duplication and only having one call to dispose?

Reply:
Yes. I have regrouped all calls to dispose in a finally block. Way better!